### PR TITLE
Native Integration for Running Unit Tests in CodeSandbox

### PR DESCRIPTION
**Pull request by gautamarora**
**What kind of change does this PR introduce?**
This PR adds native integration of unit testing using jest packages as discussed [in this issue](https://github.com/CompuIves/codesandbox-client/issues/364)

**What is the current behavior?**
Currently, there is no support for unit testing.

**What is the new behavior?**
For a project with unit tests, Code Sandbox will run those tests and show the results in a new Tests tab (next to the Console tab)

**Feature Checklist (v1)**:
- [x] Basic Unit testing
    - [x] test, it
    - [x] describe
- [x] Component testing
    - [x] Shallow rendering
    - [x] Mock functions
- [x] Auto test detection
- [x] Typescript support
- [x] Test Summary reporting

**Checklist**:

- [ ] Documentation
- [ ] Tests
- [x] Ready to be merged 
- [x] Added myself to contributors table

### CODE REVIEW

**Comment by gautamarora** [comment]
👋 @CompuIves - as you mentioned [here](https://github.com/CompuIves/codesandbox-client/issues/364#issuecomment-350134026)

> But we should add functionality to specify globals. I can do this, so you can specify an object like this:
```
{
  it,
  expect,
  // etc
}
```

could you help add `it`, `expect`, `describe`, `jestMock` from [runner](https://github.com/gautamarora/codesandbox-client/blob/f8ff153f73113e1d435482c88895665bd04b746f/packages/app/src/sandbox/eval/testers/runner.js#L22) to globals

File: packages/app/src/sandbox/compile.js [hunk 1]
```diff
@@ -192,6 +192,38 @@ async function compile({
...
+    debug(`Test Transpilation time ${Date.now() - ttt}ms`);
+
+    console.log('Tests: Running tests...');
+    //TODO need to introduce describe, it, test, jestMockFn in global scope
+    testsToTranspile.forEach(t => {
+      const evalledTests = manager.evaluateModule(t);
+    });
+
+    debug(`Test Evaluation time: ${Date.now() - tt}ms`);
+    dispatch({
```
Comment (gautamarora) [inline packages/app/src/sandbox/compile.js]: @CompuIves - the messages being dispatched from here do not seem to be showing up for me in the `handleMessage` of `Console/index.js`

**Comment by CompuIves** [comment]
Great! I've added the globals 😄 . Implementation looks really good so far.

**Comment by gautamarora** [comment]
@CompuIves - I've created a new TestManger and hooked it up to the Test Runner so that we have a nice api to run tests from compile...

```
let testManager = manager.testManager;
testManager.findTests(modules);
await testManager.transpileTests();
testManager.runTests();
let testResults = testManager.reportResults();
console.log(testResults);
```

Next, I will look into why the dispatch results are not making their way to the Console.

**Comment by gautamarora** [comment]
dispatching results to Console now works!

**Comment by gautamarora** [comment]
@CompuIves - Thank You! 
Yes, I could use your help for the Tests tab, both in adding one and thinking through how it should render the test output. One of the amazing things about the Code Sandbox experience is how clean and intuitive the UI is. I was thinking that the Tests tab should reflect that as well. Maybe we think of the Tests tab as a beautiful intuitive & helpful test reporter. Please go ahead and spin your magic! 😄

Here are 2 test sandboxes that you can try out:
Simple sandbox with 1 file and 1 test: http://pr445.cs.lbogdan.tk/s/ll4nxm6olq
React Redux TodoMVC App (all tests pass!): http://pr445.cs.lbogdan.tk/s/github/reactjs/redux/tree/master/examples/todomvc

**Comment by CompuIves** [comment]
Maan this is so impressive!

I'll try my hand at the tab, it would be great if every test result had 2 extra properties: a property with an array called something like `describe` that specifies which `describe`s it is in and a property called `file` with the file name. I want to try something with grouping the tests in the devtools.

**Comment by gautamarora** [comment]
@CompuIves - yes totally agree with grouping of tests using describe (right now i ignore `describe` but i can fix that easily) and with adding file names (not 100% sure how to do that, but will find a way).

**Comment by gautamarora** [comment]
@CompuIves - looks like all of the v1 functionality work for this is done. I will rebase this soon.

When you merge this to master, will you also be releasing a new version of the site, or is that only going to happen with v2.5 later? If this is not going to be released after merging, then I can create another PR later to add test cases for the test runner.

In the meanwhile, let me know if you have any feedback on the code.

**Comment by CompuIves** [comment]
> When you merge this to master, will you also be releasing a new version of the site, or is that only going to happen with v2.5 later? If this is not going to be released immediately after merging, then I can create another PR later to add test cases for the test runner.

It will be immediately released. I'll do a review now!

File: packages/app/src/sandbox/eval/tests/jest-lite.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+          message += `${describe} > `;
+        }
+        message += `${name}`;
+        failedMessages.push(message);
+      }
+    });
+
+    let summaryEmoji =
+      aggregatedResults.totalTestSuites === aggregatedResults.passedTestSuites
+        ? '😎'
```
Comment (CompuIves) [inline packages/app/src/sandbox/eval/tests/jest-lite.js]: Hahaha I love this

File: packages/app/src/sandbox/eval/tests/jest-lite.js~HEAD [hunk 3]
```diff
@@ -0,0 +1,210 @@
+// @flow
```
Comment (CompuIves) [inline packages/app/src/sandbox/eval/tests/jest-lite.js~HEAD]: What's with this file?

File: packages/app/src/sandbox/eval/transpiled-module.js [hunk 4]
```diff
@@ -661,7 +663,8 @@ export default class TranspiledModule {
         this.source.compiledCode,
         require,
         this.compilation,
-        manager.envVariables
+        manager.envVariables,
+        TestRunner.testGlobals()
```
Comment (CompuIves) [inline packages/app/src/sandbox/eval/transpiled-module.js]: Great!

File: packages/app/src/sandbox/eval/tests/jest-lite.js~HEAD (hunk 3 above)
Comment (gautamarora) [reply packages/app/src/sandbox/eval/tests/jest-lite.js~HEAD]: yeah, not sure. checking.

**Comment by gautamarora** [comment]
@CompuIves - i removed the console logs and that odd file.

**Comment by gautamarora** [comment]
@CompuIves - what sort of documentation (if any for now) should I add. I can add Tests for this over the next few days if you are ok merging without them.

**Comment by CompuIves** [comment]
It would be great to have documentation ready for the 25th, you can create docs here: https://github.com/CompuIves/codesandbox-client/blob/master/packages/homepage/content/docs/. In the meantime we can merge it in 😄 .

**Comment by CompuIves** [comment]
Hmm, something strange happens with the jest test, I'm not sure why...

**Pull request merged**